nemo/collections/asr/models/asrconvctcmodel.py

```diff
@@ -0,0 +1,297 @@
+        with self.__evaluation_neural_graph:
+            # Copy one input port definitions - using "user" port names.
+            self.__evaluation_neural_graph.inputs["input_signal"] = preprocessor.input_ports["input_signal"]
+            self.__evaluation_neural_graph.inputs["length"] = preprocessor.input_ports["length"]
+            # Bind the selected inputs. Connect the modules
+            i_processed_signal, i_processed_signal_len = preprocessor(
+                input_signal=self.__evaluation_neural_graph.inputs["input_signal"],
+                length=self.__evaluation_neural_graph.inputs["length"],
+            )
+            # Notice lack of speck augmentation for inference
+            i_encoded, i_encoded_len = encoder(audio_signal=i_processed_signal, length=i_processed_signal_len)
+            i_log_probs = decoder(encoder_output=i_encoded)
+            # Bind the selected outputs.
+            self.__evaluation_neural_graph.outputs["log_probs"] = i_log_probs
+            self.__evaluation_neural_graph.outputs["encoded_len"] = i_encoded_len
+
+    def __instantiate_modules(
+        self, preprocessor_params, encoder_params, decoder_params, spec_augment_params=None,
+    ):
+        preprocessor = NeuralModule.deserialize(preprocessor_params)
+        encoder = NeuralModule.deserialize(encoder_params)
+        decoder = NeuralModule.deserialize(decoder_params)
+        if hasattr(decoder, 'vocabulary'):
+            self.__vocabulary = decoder.vocabulary
+        else:
+            self.__vocabulary = None
+
+        if spec_augment_params is not None:
+            spec_augmentation = NeuralModule.deserialize(spec_augment_params)
+        else:
+            spec_augmentation = None
+
+        # Record all modules
+        self._modules = []
+        self._preprocessor = preprocessor
+        self._spec_augmentation = spec_augmentation
+        self._encoder = encoder
+        self._decoder = decoder
+        if spec_augmentation is not None:
+            self._modules += [preprocessor, spec_augmentation, encoder, decoder]
+        else:
+            self._modules += [preprocessor, encoder, decoder]
+
+        # Create input and output ports
+        self._input_ports = preprocessor.input_ports
+        self._output_ports = decoder.output_ports
+        self._output_ports['encoded_lengths'] = encoder.output_ports['encoded_lengths']
+        return self._preprocessor, self._spec_augmentation, self._encoder, self._decoder
+
+    @property
+    def train_graph(self) -> NeuralGraph:
+        return self.__training_neural_graph
+
+    @property
+    def eval_graph(self) -> NeuralGraph:
+        return self.__evaluation_neural_graph
+
+    @property
+    @add_port_docs()
+    def input_ports(self) -> Optional[Dict[str, NeuralType]]:
+        return self._input_ports
+
+    @property
+    @add_port_docs()
+    def output_ports(self) -> Optional[Dict[str, NeuralType]]:
+        return self._output_ports
```

Comment:
Maybe I should have asked https://github.com/NVIDIA/NeMo/pull/571/files?file-filters%5B%5D=.ipynb&file-filters%5B%5D=.py&file-filters%5B%5D=.yaml#r430716206 here?
How do you handle the case when eval graph and training graph have difference input and output ports?

Reply:
The way it should be handled is simple: inside of module/model input/output_ports() method(s) you write a case:
if self._operation_mode == training:
  return a
else if self._operation_mode == eval: 
 return b

Hmmm, this of course raises concerns what about "both"... I guess it should raise an error, as clearly this module/model cannot be inserted into a "both" graph, as it's inputs/outputs vary...

Reply:
>>"How do you handle the case when eval graph and training graph have difference input "
__call__ takes *kwargs and hands it to appropriate graph based on self._operation_mode

Reply:
>> "Why do NeMoModels not have predefined input and output ports?

Because they can depend on the mode